doc/nrf/ug_contribution.rst

```diff
@@ -1,126 +1,142 @@
 
-     git remote rename origin upstream
+         cd ncs/nrf
 
-   Let Git know about the fork you just created, naming it ``origin``::
+      Rename the default remote that points to the `upstream nRF Connect SDK repository <fw-nrfconnect-nrf_>`_ from ``origin`` to ``ncs``::
 
-     git remote add origin https://github.com/<your github id>/fw-nrfconnect-nrf
+         git remote rename origin ncs
 
-   and verify the remote repos::
+      Let Git know about the fork that you just created and name it ``origin``.
+      Replace *your_github_id* with your GitHub user name.
 
-     git remote -v
+      .. parsed-literal::
+         :class: highlight
 
-   The output should look similar to::
+         git remote add origin https\://github.com/*your_github_id*/fw-nrfconnect-nrf
 
-     origin   https://github.com/<your github id>/fw-nrfconnect-nrf (fetch)
-     origin   https://github.com/<your github id>/fw-nrfconnect-nrf (push)
-     upstream https://github.com/NordicPlayground/fw-nrfconnect-nrf (fetch)
-     upstream https://github.com/NordicPlayground/fw-nrfconnect-nrf (push)
+      Finally, verify the remote repositories::
 
-#. Create a topic branch (off of master) for your work (if you're addressing
-   an issue, we suggest including the issue number in the branch name)::
+         git remote -v
 
-     git checkout -b fix_comment_typo upstream/master
+      The output should look similar to this:
 
+      .. parsed-literal::
+         :class: highlight
 
-#. Make changes, test locally, change, test, test again, ...  (Check out the
-   prior chapter on `sanitycheck`_ as well).
+         origin   https\://github.com/*your_github_id*/fw-nrfconnect-nrf (fetch)
+         origin   https\://github.com/*your_github_id*/fw-nrfconnect-nrf (push)
+         ncs      https\://github.com/NordicPlayground/fw-nrfconnect-nrf (fetch)
+         ncs      https\://github.com/NordicPlayground/fw-nrfconnect-nrf (push)
 
-#. When things look good, start the pull request process by adding your changed
-   files::
+#. Create a topic branch (based on the master branch) for your work.
+   Replace *branch_name* with the name that you want to use for the local branch.
 
-     git add [file(s) that changed, add -p if you want to be more specific]
+   .. parsed-literal::
+      :class: highlight
 
-   You can see files that are not yet staged using::
+      git checkout -b *branch_name* ncs/master
+
+   If you are addressing a specific issue, we suggest including the issue number in the branch name.
+
+#. Make changes, test locally, change, test, test again, ...
+   Check out the *sanitycheck* section in Zephyr's :ref:`zephyr:Contribution Tools` documentation as well.
+
+#. When things look good, start the pull request process by adding your changed files.
+   Replace *files* with the file paths of the files that you want to add (see the Git documentation for other options).
+
+   .. parsed-literal::
+      :class: highlight
+
+      git add *files*
```

Comment:
This patch has removed the note about using `git add -p` and advocates a single `git add`.

This is in direct conflict with the guidance `As a general practice, we encourage to make small, controlled changes`. Here, "small, controlled changes" means a series of commits which each "do one thing" instead of a single git add + git commit which mixes together diff from several logically distinct changes to the source code.

The latter style of development is harder to review, and harder to bisect.

I think this documentation should continue to describe (or at least hint as it's doing now) that you can use things like `git add -p` or other tools which can stage individual hunks to split the diff into logical commits.

Here is the relevant part of the Linux documentation about contributions which spells this out in more detail:

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#separate-your-changes

Reply:
Remember that this guide is aimed at beginners, so we should describe the easiest way to do something, and not give alternatives (because people won't know which is the best alternative for them to use).
I've never used ``add -p`` - is it easier than just listing the files (if you put yourself in the shoes of someone who is NOT familiar with git)? I tried it out now and it ignored untracked files.

Reply:
> Remember that this guide is aimed at beginners, so we should describe the easiest way to do something, and not give alternatives (because people won't know which is the best alternative for them to use).

In my personal opinion, if you are going to contribute to NCS, knowledge of how to structure your git commits for ease of review is an important skill, and we should expect that level of professionalism from potential contributors.

This isn't about separating changes from separate files into separate commits, it's about setting up your commits so they each do one logical thing. Using `git add -p` is useful if your working tree contains changes in one file that should logically be part of two distinct commits, because it allows you to stage individual hunks from each file you're adding to the index. There are other ways to accomplish the goal -- as this patch itself said, " As usual with Git, there are multiple ways of getting a task done" -- but this is a common recommendation since it's built into git.

@carlescufi do you have any thoughts on this?